### PR TITLE
Add Item#canBeStoredInItemInventory (1.12)

### PR DESCRIPTION
**Pull request by malte0811**
See #3954 for details.
@mezz do you want me to push commits to the 1.11 version whenever I push them here, when this PR is accepted or not at all?

### CODE REVIEW

**Comment by mezz** [comment]
It's probably easiest if you wait for this to be merged, then port it to 1.11 any time after.

**Comment by gigaherz** [comment]
Isn't Forge's stance that `IInventory` doesn't exist except as an implementation detail? I mean, shouldn't this code work with `IItemHandler` exclusively, and make use the existing wrappers for vanilla things?

**Comment by malte0811** [comment]
I don't have the code here right now, but aren't `IInventory`'s still officially used for GUI/container things? @mezz does not really like the idea of `PortableInventoryLocation`'s, so I will probably remove them anyways.

**Comment by gigaherz** [comment]
We have `SlotItemHandler` for using the `IItemHandler` in `Container`. The vanilla code does use `IInventory`+`Slot`, but so far as I know, the official stance is that mods shouldn't do that, and should use capabilities instead.

**Comment by malte0811** [comment]
If that is the case I will remove the IInventory from the inventory location class.
Is there some sort of list of these "official stances" somewhere? The more I talk to people in this repo and on the Forge IRC channel the more I feel like I am doing everything wrong (even though it works perfectly fine).

**Comment by mezz** [comment]
You'll get any relevant "official stances" from the horse's mouth generally. 
Lots of people help out with PRs by giving comments and suggestions, but they are only merged by lex, cpw, fry, or me. Usually I review and then hand it off to one of the others for a final review before merging.

I think you should either remove `PortableInventoryLocation` or brainstorm a different solution that conveys more information without requiring the consume to do `instanceof` checks.
Once you've addressed that I will mark this for review from lex.
Hope that clarifies the process a bit.

**Comment by malte0811** [comment]
I wasn't talking about the PR's but about my own mods. I usually look at either vanilla or IE's code for references since I have those in my workspaces, but neither of those follows these best practices.
I still feel like there should be some way for items to know what they are being put into, I can't think of any better way to achieve that though. As I said before I won't do any work on this PR for the next ~2 weeks, partially because I want to take a break from modding, partially because genPatches would probably take an hour to complete on my laptop.

**Comment by mezz** [comment]
Maybe you can just pass the container as an `Object` parameter?
It still has the same issues as `PortableInventoryLocation` but is simpler.

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 1]
```diff
@@ -1,10 +1,59 @@
...
+@@ -58,10 +34,16 @@
  import net.minecraftforge.fml.relauncher.Side;
  import net.minecraftforge.fml.relauncher.SideOnly;
  
 -public class Item
++import javax.annotation.Nullable;
++import java.util.List;
++import java.util.Map;
++import java.util.Random;
++import java.util.UUID;
```
Comment (covers1624) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: Don't change imports in patches.

Comment (malte0811) [reply patches/minecraft/net/minecraft/item/Item.java.patch]: I know, I forgot to check before pushing (I was in a hurry). I'll fix those imports tomorrow.

**Comment by howtonotwin** [comment]
That string-typing makes me feel icky. The only ~~two~~ three (see below) sane ways I can think of doing this is a `Map<ResourceLocation, Object>` (not `String`, because rlocs need to be namespaced), and the old way you had it, where the latter seems preferable to me. Maybe instead of using classes, use a hierarchy of interfaces, so mods can have e.g `ItemContainer & SomethingElseContainer`. There's really no way to get around the inherent type-unsafety here, so we may as well go for *nice* type-unsafety. <sup><sub>/me wonders if someone has implemented <code>HMap</code> in Java</sub></sup>

EDIT: Yep, you can do a neutered form of `HMap` in Java.
We could *try* something like

```java
// These HMaps are transformations from the Key functor to the Id functor.
// Not very generic, but it suffices for a good many usecases.
@SupressWarnings("unchecked")
public final class HMap
{
    private final Map<Object, Object> underlying = new HashMap<>();
    @Nullable
    public <T> T get(Key<T> key)
    {
        // EDIT: See below
            if(key == null) throw new NullPointerException("Tried to get null key");
        // END EDIT
        return (T)underlying.get(key);
    }
    @Nullable
    public <T> T put(Key<T> key, T value)
    {
        // EDIT: See below
            if(key == null) throw new NullPointerException("Tried to put null key");
        // END EDIT
        return (T)underlying.put(key, value);
    }
    public static final class Key<T> {}
}
```

And then we could stick some `public static final` `Key`s somewhere, like `Key<InvType> INVENTORY_TYPE` and mods can safely do `Key<Integer> COMPRESSION_LEVEL`, and then call `hmap.get(COMPRESSION_LEVEL)`, and that guarantees that a) it is the correct type and b) no one else has managed to collide with the key, because the keys use identity equality. But I don't think any other place in Forge that has this problem uses this approach, so there's that.

**Comment by howtonotwin** [comment]
Oh, and for future reference, git commit messages should be in the imperative mood, i.e. they should sound like commands: "Fix the imports in Item.java". (No need to change anything, just a note.) [Here's a nice reference](https://chris.beams.io/posts/git-commit/)

**Comment by malte0811** [comment]
Re commit messages: I always thought of commit messages as description of what I had done, rather than descriptions of what the commit does. I really don't think it matters that much, it is just a minor grammatical difference. The content stays the same.

Re the actual topic of this PR:

@mezz does not like the way I had it before, so that is not really an option.

Using a Map could work too, I'd like to know @mezz's opinion on that though. The main advantage of the `HMap`-class you showed would be the ability to store any type without the chance of mixing up types, is that correct? (The only reference to `HMap` I found was in Ruby and I don't know ANY Ruby)
I have a third idea for a Map-based approach, a mix of the 2 you suggested:
I like the idea of "implicit" type safety while still being able to store any `Object` (The way you edited in). I don't really like the idea of having to add a mod as a compile-time dependency just because one needs to be able to access the map key (And one would have to wrap any `hasKey` checks in `isModLoaded` checks). As an example IC2 might add a key storing a `boolean` value indicating radiation proof storage items. I'd use something like [this](https://gist.github.com/malte0811/e598861e30ce6b0cbbe1bfa81f1291a9) as a key: Mods could create keys for properties added by other mods without those being present as long as the type is available.

**Comment by howtonotwin** [comment]
Re: commit messages: Git creates commits in the imperative whenever you merge/cherrypick/etc., so any human made commit should be the same.

Re: the map: I don't really like having the keys be any more complicated than simple dummies. It makes perfect sense to need a compile dep on a mod in order to interop with it (at least, if that interop is more complicated than plain IMC). Anyway, I think mods should store an `Optional<Key<T>>` in some field depending on whether a mod is loaded, and then they just say `optKey.map(hmap::get)`. Also, the above impl of `HMap` is actually unsafe.  Because `null` is `Key<T>` for all `T`, you can put a `String` at `null` and then try to get an `Integer`. It should be modified to throw on `null` keys.

EDIT: `Optional` won't protect against not actually having the *type* available, but it's safe to have fields that hold `Optional<Key<AbsentT>>` because of erasure, and then logic can be protected by saying `if(optKey.isPresent()) SomeHelperClass.compute(hmap, opt.get())`, where `SomeHelperClass` is only loaded if the other mod is loaded, as you would in other cases where types may not exist.

**Comment by malte0811** [comment]
I didn't think about using `Optional`, I should use that class more often in general.
I didn't want to have more compile-time dependencies because the last time I set dependencies as `compileOnly` in gradle IntelliJ just ignored that (and launching a client and a server with a lot of mods in 1.8+ while IntelliJ is running with "just" 8GB of RAM is not fun). It works now, I don't know what I did different. So your `HMap` method should be fine assuming @mezz approves using it.
I'd say it needs a better name though, when you asked whether anyone had implemented a `HMap` in Java yesterday it confused me quite a bit since I thought `HMap` was just your personal abbreviation for `HashMap` (which someone obviously implemented in Java).

**Comment by howtonotwin** [comment]
`HMap` means heterogeneous map. In the most general case (see shapeless's `HMap`, written in Scala (shapeless is a library that introduces massive amounts of type-level programming into Scala)), they enforce some arbitrary relation between key types and value types. For example, using shapeless's `HMap`, I could enforce that all `Nat`ural number keys go to a list of `String`s of the same size (using another type from shapeless, `Sized`). However, this requires language support for higher kinds, which Java does not have. There must exist some way to emulate it in Java, because our type system still manages Turing completeness, but it would likely force us to encode all the type computations in a complicated inheritance hierarchy. The above implementation is a restriction of the truly general `HMap` to the transformation `Key<T> => T`. I'd just call it `HMapKeyToId` (as `Id<T> = T`), but that's not very descriptive.

[Paper on Java's Turing completeness](https://arxiv.org/pdf/1605.05274)

**Comment by mezz** [comment]
The path you are investigating is breaking OO patterns badly...
If this logic really depends on the other container, can we make it the one that decides what can go in instead?

**Comment by howtonotwin** [comment]
The way I see it, the container needs to be able to validate the items placed in it somehow, which is already done simply by how `IItemHandler` works, and we also need the the item to validate the container, so we need to ask whether the item wants to go in a container. The problem is in getting information between the container and item. A container normally validates items by checking the kind of item it is, and it often uses capabilities to check extra information.

Dualizing, items, if they are going to do a similar kind of validation, will need to check what kind of `IItemHandler` their containers are, and they will often need to use "capabilities" to check extra information. This would require allowing `IItemHandler`s themselves to have capabilities, or something like them. Looking back at it, capabilities actually work a whole lot like `HMap`, so I don't understand why you say this is breaking patterns when it is already a big part of Forge.

So, the suggestion of a `Map<String/ResourceLocation, Object>` is disqualified simply because `HMap` is the same thing but better, and `HMap` is close to `Capability`, so we consider them a single alternative. The other alternative we have is the same thing we did before `Capability`, which is a bunch of interfaces, and which was incarnated in a slightly less flexible manner using classes in the original PR.

**Comment by mezz** [comment]
You either need to reduce the container to a set of properties you care about, which any item can understand, or you need to have the container do the logic. There are thousands of mods, you can't check for each one that might be interesting to your item. Strings and maps of arbitrary properties are not much better.

**Pull request closed without merging**